File: Python3/Hard/Accepted/largest-rectangle-in-histogram_5.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        
        if not heights:
            return 0
        stack = []
        heights.append(0)
        max_area = 0
        for i in range(len(heights)):
            while stack and heights[i] < heights[stack[-1]]:
                h = heights[stack.pop()]
                w = i if not stack else i - stack[-1] - 1
                max_area = max(max_area, h * w)
            stack.append(i)
        return max_area

File: Rust/Hard/CompileError/largest-rectangle-in-histogram_5.rs
impl Solution {
    pub fn largest_rectangle_area(heights: Vec<i32>) -> i32 {
        
        let mut stack = Vec::new();
        let mut max_area = 0;
        let mut i = 0;
        while i < heights.len() {
            if stack.is_empty() || heights[i] > heights[*stack.last().unwrap()] {
                stack.push(i);
                i += 1;
            } else {
                let top = stack.pop().unwrap();
                let area = heights[top] * (if stack.is_empty() { i as i32 } else { i as i32 - stack.last().unwrap() as i32 - 1 });
                max_area = max_area.max(area);
            }
        }
        while !stack.is_empty() {
            let top = stack.pop().unwrap();
            let area = heights[top] * (if stack.is_empty() { i as i32 } else { i as i32 - stack.last().unwrap() as i32 - 1 });
            max_area = max_area.max(area);
        }
        max_area
    }
}